rfc/005-soft-chain-upgrades.md

```diff
@@ -0,0 +1,438 @@
+
+As we expect upgrades to always increment the block version by one, rather
+than having the version number being passed in `EndBlockResponse{}`, another
+option could be to have the height where we want the upgrade to occur. With this
+approach validators are not reaching consensus on the version but the window of
+time (as a height) that nodes will have to perform or prepare for the upgrade.
+
+We will now cover the two main methods of executing a soft upgrade.
+
+## Method 1: Multi Data Structure Support
+
+This would require Tendermint to support all prior block protocol version. This
+would be done by somehow finding what the version of the structure is and
+processing it accordingly. This mimics a similar design model to the Cosmos SDK
+where all legacy code is stored in the repository and conditionals are used to
+indicate how messages are processed.
+
+One could imagine a directory structure for each module as
+such:
+
+```go
+consensus // module
+|---- state.go
+|---- state_test.go
+|       ...
+|---- votes.go // current version (v10)
+|       ...
+|---- legacy
+   |---- v8 // prior supported versions
+      |---- votes.go
+      |      ...    // other data structures ( and pure functions )
+   |---- v9
+      |---- votes.go
+      |      ...
+```
+
+The module, depending on the types, may statically cast each of the different
+versions or define an interface that it expects all versions to follow. For
+example:
+
+```go
+func (blockExec *BlockExecutor) CreateProposalBlock(state State) (BlockI, error)
+{
+ switch state.Version.Block {
+ case 9:
+  return v9.CreateProposalBlock(state, blockExec.Mempool, blockExec.Evpool)
+ case 10:
+  return v10.CreateProposalBlock(state, blockExec.Mempool, blockExec.Evpool)
+ default:
+  return fmt.Errorf("block version %d not recognised", state.Version.Block)
+ }
+}
+```
+
+In this case, Block which falls under a data structure of the block protocol
+could have many versions and is represented here as an interface. `State` and
+`BlockExecutor` are structs outside of the block protocol so would be
+internalized to allow for changes (so long as this wouldn't change the overall
+behavior).
+
+### Current Use of Versioning
+
+Records of versions are currently stored in `State` and in the `Header` of each
+block. This already covers a broad base of scenarios. The consensus reactor can
+use the version from state to ensure it reads `Vote`s and `Partsetheader`s
+correctly. Similarly, fast sync, the block executor and evidence module, whom
+all interact with block protocol data structures also all have access to state.
+The next few sections will explore the aspects of Tendermint that will need to
+be modified to prepare for multi versioning support.
+
+### RPC
+
+Supporting soft upgrades means that the RPC will need to be able to deliver
+information across different versions. The RPC will therefore need to know which
+heights corresponds to which version and then be able to communicate this to
```

Comment:
Note that the data structure version of the rpc query depends on the content of the query, i.e, height. If I am not wrong, then either rpc queries will be two step (make a query, full node responds with a right version (if wrong), then make a second query using right version), or rpc client will need to have logic conditional on the data itself.

Reply:
We expect, with this method, that the rpc client will have the conditional logic to process the data structure correctly. This means it should support all the versions that the full node supports. the version number will be put in the call i.e. /v3/block?height=100 where the version is the latest that rpc client supports. If it is insufficient then the full node will return some versioning error else it will return the data structure and the version (which could be v1 or v2)

Reply:
I will try make this clearer in the section here